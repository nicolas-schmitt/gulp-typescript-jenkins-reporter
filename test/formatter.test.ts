import { flatten, forEach } from 'lodash';
import Formatter from '../src/formatter';
import inputMock from './mocks/input';
import outputMock from './mocks/output';
import { parseAndValidate } from './util';
import xpath from 'xpath';

describe('formatter', () => {
    describe('formatError', () => {
        it('should return valid xml', () => {
            forEach(inputMock, input => {
                input.errors.forEach(error => {
                    const formatter = new Formatter();
                    parseAndValidate(formatter.formatError(error as any));
                });
            });
        });

        it('should return the xml corresponding to the error', () => {
            forEach(inputMock, (input, name) => {
                input.errors.forEach((error, i) => {
                    const formatter = new Formatter();
                    const actualXml = formatter.formatError(error as any);
                    expect(actualXml).toEqual(outputMock[name as keyof typeof inputMock].errors[i]);
                });
            });
        });
    });

    describe('formatFile', () => {
        it('should return valid xml', () => {
            forEach(outputMock, mock => {
                const formatter = new Formatter();
                const errors = mock.errors.map(error => ({
                    path: mock.fullFilename,
                    xml: error,
                }));

                const actualXml = formatter.formatFile(mock.fullFilename, errors);

                if (errors.length) {
                    parseAndValidate(actualXml);
                } else {
                    expect(actualXml).toBe('');
                }
            });
        });

        it('should return an empty string when there is no error', () => {
            const formatter = new Formatter();
            expect(formatter.formatFile(outputMock.clean.fullFilename, [])).toBe('');
        });

        it('should return a file element with as many children as failure', () => {
            forEach(outputMock, (mock, name) => {
                const formatter = new Formatter();
                const errors = mock.errors.map(error => ({
                    path: mock.fullFilename,
                    xml: error,
                }));

                const xml = formatter.formatFile(mock.fullFilename, errors);
                if (name === 'clean') {
                    expect(xml).toBe('');
                } else {
                    const doc = parseAndValidate(xml);
                    expect(xpath.select('//violation', doc).length).toBe(errors.length);
                }
            });
        });
    });

    describe('formatStream', () => {
        const errors = flatten(
            Object.entries(outputMock).map(([_, mock]) =>
                mock.errors.map(error => ({
                    path: mock.relativeFilename,
                    xml: error,
                }))
            )
        );

        it('should return a valid xml', () => {
            const formatter = new Formatter();

            parseAndValidate(formatter.formatStream(errors));
        });

        it('should call #formatFile for each file with error', () => {
            const spy = jest.spyOn(Formatter.prototype, 'formatFile');
            const formatter = new Formatter();
            formatter.formatStream(errors);

            expect(spy).toHaveBeenCalledTimes(2);
        });

        it('should return a pmd element with as many children as there are files with errors', () => {
            const formatter = new Formatter();
            const doc = parseAndValidate(formatter.formatStream(errors));
            expect(xpath.select('//pmd', doc).length).toBe(1);
            expect(xpath.select('//file', doc).length).toBe(2);
        });
    });
});
