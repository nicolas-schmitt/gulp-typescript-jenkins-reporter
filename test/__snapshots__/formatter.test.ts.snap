// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`formatter formatError returns the xml corresponding to the error awful[0] 1`] = `
<violation
  begincolumn="22"
  beginline="1"
  endcolumn="35"
  endline="1"
  priority="1"
  rule="TS2307"
>
  Cannot find module './dirtyy.ts'.
</violation>
`;

exports[`formatter formatError returns the xml corresponding to the error awful[1] 1`] = `
<violation
  begincolumn="14"
  beginline="7"
  endcolumn="23"
  endline="7"
  priority="1"
  rule="TS2339"
>
  Property 'something' does not exist on type 'Awful'.
</violation>
`;

exports[`formatter formatError returns the xml corresponding to the error awful[2] 1`] = `
<violation
  begincolumn="13"
  beginline="11"
  endcolumn="16"
  endline="11"
  priority="1"
  rule="TS7005"
>
  Variable 'tab' implicitly has an 'any[]' type.
</violation>
`;

exports[`formatter formatError returns the xml corresponding to the error awful[3] 1`] = `
<violation
  begincolumn="25"
  beginline="16"
  endcolumn="30"
  endline="16"
  priority="1"
  rule="TS2304"
>
  Cannot find name 'Dirty'.
</violation>
`;

exports[`formatter formatError returns the xml corresponding to the error awful[4] 1`] = `
<violation
  begincolumn="10"
  beginline="27"
  endcolumn="11"
  endline="27"
  priority="1"
  rule="TS2430"
>
  Interface 'B' incorrectly extends interface 'A'.
  Types of property 'prop' are incompatible.
    Type 'string' is not assignable to type 'number'.
</violation>
`;

exports[`formatter formatError returns the xml corresponding to the error dirty[0] 1`] = `
<violation
  begincolumn="17"
  beginline="6"
  endcolumn="22"
  endline="6"
  priority="1"
  rule="TS2339"
>
  Property 'loggg' does not exist on type 'Console'.
</violation>
`;

exports[`formatter formatError returns the xml corresponding to the error dirty[1] 1`] = `
<violation
  begincolumn="16"
  beginline="11"
  endcolumn="22"
  endline="11"
  priority="1"
  rule="TS2322"
>
  Type 'string' is not assignable to type 'number'.
</violation>
`;

exports[`formatter formatFile returns a file element with violations for awful 1`] = `
<file
  name="/home/ubuntu/workspace/sandbox/data/awful.ts"
>
  <violation
    begincolumn="22"
    beginline="1"
    endcolumn="35"
    endline="1"
    priority="1"
    rule="TS2307"
  >
    Cannot find module './dirtyy.ts'.
  </violation>
  <violation
    begincolumn="14"
    beginline="7"
    endcolumn="23"
    endline="7"
    priority="1"
    rule="TS2339"
  >
    Property 'something' does not exist on type 'Awful'.
  </violation>
  <violation
    begincolumn="13"
    beginline="11"
    endcolumn="16"
    endline="11"
    priority="1"
    rule="TS7005"
  >
    Variable 'tab' implicitly has an 'any[]' type.
  </violation>
  <violation
    begincolumn="25"
    beginline="16"
    endcolumn="30"
    endline="16"
    priority="1"
    rule="TS2304"
  >
    Cannot find name 'Dirty'.
  </violation>
  <violation
    begincolumn="10"
    beginline="27"
    endcolumn="11"
    endline="27"
    priority="1"
    rule="TS2430"
  >
    Interface 'B' incorrectly extends interface 'A'.
  Types of property 'prop' are incompatible.
    Type 'string' is not assignable to type 'number'.
  </violation>
</file>
`;

exports[`formatter formatFile returns a file element with violations for dirty 1`] = `
<file
  name="/home/ubuntu/workspace/sandbox/data/dirty.ts"
>
  <violation
    begincolumn="17"
    beginline="6"
    endcolumn="22"
    endline="6"
    priority="1"
    rule="TS2339"
  >
    Property 'loggg' does not exist on type 'Console'.
  </violation>
  <violation
    begincolumn="16"
    beginline="11"
    endcolumn="22"
    endline="11"
    priority="1"
    rule="TS2322"
  >
    Type 'string' is not assignable to type 'number'.
  </violation>
</file>
`;

exports[`formatter formatFile returns no file element with violations for clean 1`] = `undefined`;

exports[`formatter formatStream returns a pmd element with file elements as children 1`] = `
<pmd
  version="4.2.5"
>
  <file
    name="data/dirty.ts"
  >
    <violation
      begincolumn="17"
      beginline="6"
      endcolumn="22"
      endline="6"
      priority="1"
      rule="TS2339"
    >
      Property 'loggg' does not exist on type 'Console'.
    </violation>
    <violation
      begincolumn="16"
      beginline="11"
      endcolumn="22"
      endline="11"
      priority="1"
      rule="TS2322"
    >
      Type 'string' is not assignable to type 'number'.
    </violation>
  </file>
  <file
    name="data/awful.ts"
  >
    <violation
      begincolumn="22"
      beginline="1"
      endcolumn="35"
      endline="1"
      priority="1"
      rule="TS2307"
    >
      Cannot find module './dirtyy.ts'.
    </violation>
    <violation
      begincolumn="14"
      beginline="7"
      endcolumn="23"
      endline="7"
      priority="1"
      rule="TS2339"
    >
      Property 'something' does not exist on type 'Awful'.
    </violation>
    <violation
      begincolumn="13"
      beginline="11"
      endcolumn="16"
      endline="11"
      priority="1"
      rule="TS7005"
    >
      Variable 'tab' implicitly has an 'any[]' type.
    </violation>
    <violation
      begincolumn="25"
      beginline="16"
      endcolumn="30"
      endline="16"
      priority="1"
      rule="TS2304"
    >
      Cannot find name 'Dirty'.
    </violation>
    <violation
      begincolumn="10"
      beginline="27"
      endcolumn="11"
      endline="27"
      priority="1"
      rule="TS2430"
    >
      Interface 'B' incorrectly extends interface 'A'.
  Types of property 'prop' are incompatible.
    Type 'string' is not assignable to type 'number'.
    </violation>
  </file>
</pmd>
`;
